File: mobile/lib/data/repositories/settings_repository.dart
import 'package:shared_preferences/shared_preferences.dart';

import '../services/controller_preset.dart';
import '../services/controller_visibility.dart';
import '../services/dashboard_input.dart';
import '../services/dashboard_visibility.dart';
import '../services/engine_start_mode.dart';
import '../services/input_mapping.dart';
import '../services/pedal_input.dart';
import '../services/pedal_side.dart';
import '../services/spring_back.dart';
import '../services/wheel_mode.dart';

/// Single source of truth for settings (mapping, visibility, pedal sides,
/// wheel mode, rotation degrees, bindings).
class SettingsRepository {
  const SettingsRepository();

  Future<InputMapping> getMapping() => InputMapping.load();
  Future<void> setMapping(InputMapping mapping) => mapping.save();

  Future<ControllerVisibility> getVisibility() => ControllerVisibility.load();
  Future<void> setVisibility(ControllerVisibility v) => v.save();

  Future<GamePreset> getPreset() => GamePreset.load();
  Future<void> setPreset(GamePreset preset) => preset.save();

  Future<WheelMode> getWheelMode() => WheelMode.load();
  Future<void> setWheelMode(WheelMode mode) => mode.save();

  Future<int> getRotationDegree(GamePreset preset) =>
      RotationDegree.load(preset);
  Future<void> setRotationDegree(GamePreset preset, int degree) =>
      RotationDegree.save(preset, degree);

  Future<bool> getSpringBack() => SpringBack.load();
  Future<void> setSpringBack(bool value) => SpringBack.save(value);

  Future<PedalSides> getPedalSides() => PedalSides.load();

  Future<EngineStartMode> getEngineStartMode() => EngineStartMode.load();
  Future<void> setEngineStartMode(EngineStartMode mode) => mode.save();

  Future<DashboardVisibility> getDashboardVisibility() =>
      DashboardVisibility.load();
  Future<void> setDashboardVisibility(DashboardVisibility visibility) =>
      visibility.save();

  Future<void> setPedalSide(PedalType pedal, PedalSide side) async {
    final current = await PedalSides.load();
    await current.copyWithSide(pedal, side).save();
  }

  static String bindingKey(ControlId c, bool isGamepad) =>
      'wheeldeck.binding.${isGamepad ? 'gamepad' : 'keyboard'}.${c.wireValue}';

  Future<String?> getBindingOverride(ControlId c, bool isGamepad) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(bindingKey(c, isGamepad));
  }

  Future<void> setBindingOverride(
    ControlId c,
    bool isGamepad,
    String value,
  ) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(bindingKey(c, isGamepad), value);
  }

  Future<void> clearBindingOverrides() async {
    final prefs = await SharedPreferences.getInstance();
    for (final c in ControlId.values) {
      await prefs.remove(bindingKey(c, true));
      await prefs.remove(bindingKey(c, false));
    }
  }

  Future<void> resetAll() async {
    await InputMapping.fallback.save();
    await GamePreset.ets2.save();
    await const ControllerVisibility(
      showClutch: false,
      showDashboard: true,
    ).save();
    await PedalSides({}).save();
    await WheelMode.rotatable.save();
    await SpringBack.save(SpringBack.fallback);
    await EngineStartMode.holdConfirm.save();
    await const DashboardVisibility(DashboardVisibility.defaults).save();
    for (final preset in GamePreset.values) {
      await RotationDegree.save(preset, RotationDegree.fallback);
    }
    await clearBindingOverrides();
  }
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/repositories/SettingsRepository.kt
package dev.fazrigading.wheeldeck.data.repositories

import dev.fazrigading.wheeldeck.data.services.CameraPadMode
import dev.fazrigading.wheeldeck.data.services.ControllerVisibility
import dev.fazrigading.wheeldeck.data.services.DashboardVisibility
import dev.fazrigading.wheeldeck.data.services.EngineStartMode
import dev.fazrigading.wheeldeck.data.services.GamePreset
import dev.fazrigading.wheeldeck.data.services.InputMapping
import dev.fazrigading.wheeldeck.data.services.RotationDegree
import dev.fazrigading.wheeldeck.data.services.SettingsStore
import dev.fazrigading.wheeldeck.data.services.SpringBack
import dev.fazrigading.wheeldeck.data.services.WheelMode

/// Single source of truth for driving settings: mapping, visibility, wheel
/// mode, rotation degrees, and the rest of the persisted toggles.
///
/// Every getter applies the ported default when the key was never written, so
/// callers never see a null. Bindings, pedal sides, and reset land with Task 12.
class SettingsRepository(private val store: SettingsStore) {

    suspend fun getMapping(): InputMapping =
        InputMapping.fromWireValue(store.loadString(InputMapping.KEY))

    suspend fun setMapping(mapping: InputMapping) =
        store.saveString(InputMapping.KEY, mapping.wireValue)

    suspend fun getPreset(): GamePreset =
        GamePreset.fromWireValue(store.loadString(GamePreset.KEY))

    suspend fun setPreset(preset: GamePreset) =
        store.saveString(GamePreset.KEY, preset.wireValue)

    suspend fun getWheelMode(): WheelMode =
        WheelMode.fromWireValue(store.loadString(WheelMode.KEY))

    suspend fun setWheelMode(mode: WheelMode) =
        store.saveString(WheelMode.KEY, mode.wireValue)

    suspend fun getRotationDegree(preset: GamePreset): Int =
        RotationDegree.resolve(store.loadInt(RotationDegree.key(preset)))

    suspend fun setRotationDegree(preset: GamePreset, degree: Int) {
        require(degree in RotationDegree.allowed) { "Unsupported rotation degree: $degree" }
        store.saveInt(RotationDegree.key(preset), degree)
    }

    suspend fun getSpringBack(): Boolean = store.loadBool(SpringBack.KEY) ?: SpringBack.FALLBACK

    suspend fun setSpringBack(value: Boolean) = store.saveBool(SpringBack.KEY, value)

    suspend fun getCameraPadMode(): CameraPadMode =
        CameraPadMode.fromWireValue(store.loadString(CameraPadMode.KEY))

    suspend fun setCameraPadMode(mode: CameraPadMode) =
        store.saveString(CameraPadMode.KEY, mode.wireValue)

    suspend fun getEngineStartMode(): EngineStartMode =
        EngineStartMode.fromWireValue(store.loadString(EngineStartMode.KEY))

    suspend fun setEngineStartMode(mode: EngineStartMode) =
        store.saveString(EngineStartMode.KEY, mode.wireValue)

    suspend fun getDashboardVisibility(): DashboardVisibility =
        DashboardVisibility.fromWireValues(store.loadStringSet(DashboardVisibility.KEY))

    suspend fun setDashboardVisibility(visibility: DashboardVisibility) =
        store.saveStringSet(DashboardVisibility.KEY, visibility.wireValues())

    /// Reads the controller visibility, migrating a legacy 5-way value on the
    /// way through and removing it.
    suspend fun getVisibility(): ControllerVisibility {
        val legacy = store.loadString(ControllerVisibility.LEGACY_KEY)
        if (legacy != null) {
            val migrated = ControllerVisibility.fromLegacy(legacy)
            setVisibility(migrated)
            store.remove(ControllerVisibility.LEGACY_KEY)
            return migrated
        }
        val clutch = store.loadBool(ControllerVisibility.CLUTCH_KEY)
        val dashboard = store.loadBool(ControllerVisibility.DASHBOARD_KEY)
        if (clutch == null && dashboard == null) return ControllerVisibility.fallback
        return ControllerVisibility(
            showClutch = clutch ?: ControllerVisibility.fallback.showClutch,
            showDashboard = dashboard ?: ControllerVisibility.fallback.showDashboard,
        )
    }

    suspend fun setVisibility(visibility: ControllerVisibility) {
        store.saveBool(ControllerVisibility.CLUTCH_KEY, visibility.showClutch)
        store.saveBool(ControllerVisibility.DASHBOARD_KEY, visibility.showDashboard)
    }
}
